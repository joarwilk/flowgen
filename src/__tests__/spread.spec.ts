import { compiler, beautify } from "..";
import "../test-matchers";

it("should use spread when performing union of exact object types", () => {
  const ts = `
type Foo = { foo: number };
type Bar = { bar: string };
const combination: Foo & Bar;
`;
  const result = compiler.compileDefinitionString(ts, {
    quiet: true,
    inexact: false,
  });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should not use spread when performing union of inexact object types", () => {
  const ts = `
type Foo = { foo: number };
type Bar = { bar: string };
const combination: Foo & Bar;
`;
  const result = compiler.compileDefinitionString(ts, {
    quiet: true,
    inexact: true,
  });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should not insert spread when performing union of class types", () => {
  const ts = `
class Foo {}
class Bar {}
const combination: Foo & Bar;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
