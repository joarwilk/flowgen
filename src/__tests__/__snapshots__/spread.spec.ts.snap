// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not insert spread when performing union of class types 1`] = `
"declare class Foo {}
declare class Bar {}
declare var combination: Foo & Bar;
"
`;

exports[`should not use spread when performing union of inexact object types 1`] = `
"declare type Foo = {
  foo: number,
  ...
};
declare type Bar = {
  bar: string,
  ...
};
declare var combination: Foo & Bar;
"
`;

exports[`should use spread when performing union of exact object types 1`] = `
"declare type Foo = {|
  foo: number,
|};
declare type Bar = {|
  bar: string,
|};
declare var combination: {| ...Foo, ...Bar |};
"
`;
